org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacBindingResolver.java

```diff
@@ -150,7 +150,13 @@ public JavacPackageBinding getPackageBinding(PackageSymbol packageSymbol) {
 		}
 		//
 		private Map<String, JavacTypeBinding> typeBinding = new HashMap<>();
+		public JavacTypeBinding getTypeBinding(JCTree tree, com.sun.tools.javac.code.Type type) {
+			return getTypeBinding(type, tree instanceof JCClassDecl);
+		}
 		public JavacTypeBinding getTypeBinding(com.sun.tools.javac.code.Type type) {
+			return getTypeBinding(type, false);
```

Comment:
why does it set the last argument to false?

Reply:
The `getTypeBinding(com.sun.tools.javac.code.Type type)` entry point is exclusively used when starting with a different Symbol and somehow working yourself up the chain to a ClassType. This means that, when using this entry point, we are almost always using a symbol representing a specific instance of a class. This has repercussions for generics. There's a difference between a `ClassDeclaration` AST node and the underlying javac symbols that back it. If the class is a generic, a ClassDeclaration will typically be `X<T>` rather than `X<String>`. 

Anyway, it's important to know if we're dealing with a class declaration or not so we know how to interpret the generic parameters.  This entry is exclusively used by code that follows up a trail of Symbol owners, which means the original symbol is NOT a `ClassDeclaration` object and does not have the generic parameters necessary to make the tests pass.